django/test/utils.py

```diff
@@ -170,71 +172,140 @@ def teardown_test_environment():
+        self.verbosity = verbosity
+        self.interactive = interactive
+        self.time_keeper = time_keeper or NullTimeKeeper()
+        self.keepdb = keepdb
+        self.debug_sql = debug_sql
+        self.parallel = parallel
+        self.aliases = aliases
+        self.serialized_aliases = serialized_aliases
+        self.test_databases = {}
+        self.mirrored_aliases = {}
+        self.old_names = []
+        self.main_db_created = threading.Event()
+
+    def setup(self):
+        """Main method to set up test databases."""
+        self.test_databases, self.mirrored_aliases = get_unique_databases_and_mirrors(
+            self.aliases
+        )
+
+        # Process each unique database
+        for db_sig, (db_name, aliases) in self.test_databases.items():
+            # Extract the first alias outside the loop
+            first_alias = aliases[0]
+            remaining_aliases = aliases[1:]
 
-    # Configure the test mirrors.
-    for alias, mirror_alias in mirrored_aliases.items():
-        connections[alias].creation.set_as_test_mirror(
-            connections[mirror_alias].settings_dict
+            # Create the test database for the first alias
+            self.create_test_database(first_alias, db_name)
+
+            # Clone the test database if parallel testing is enabled
+            if self.parallel > 1:
+                self.clone_databases_in_parallel(first_alias)
+
+            # Configure remaining aliases as mirrors
+            for alias in remaining_aliases:
+                self.configure_as_mirror(alias, first_alias)
+
+        # Configure any additional test mirrors
+        self.configure_test_mirrors()
+
+        # Enable debug SQL logging if required
+        if self.debug_sql:
+            self.enable_debug_sql()
+
+        return self.old_names
+
+    def create_test_database(self, alias, db_name):
+        """Create the test database for the given alias."""
+        connection = connections[alias]
+        self.old_names.append((connection, db_name, True))
+
+        serialize_alias = (
+            self.serialized_aliases is None or alias in self.serialized_aliases
         )
 
-    if debug_sql:
+        with self.time_keeper.timed(f"  Creating '{alias}'"):
+            connection.creation.create_test_db(
+                verbosity=self.verbosity,
+                autoclobber=not self.interactive,
+                keepdb=self.keepdb,
+                serialize=serialize_alias,
+            )
+
+        self.main_db_created.set()
+
+    def clone_databases_sequentially(self, alias):
+        """Clone the test database sequentially for parallel test execution."""
+        for index in range(self.parallel):
+            self.clone_test_database(alias, index)
+
+    def clone_databases_in_parallel(self, alias):
+        """Clone the test database in parallel threads for parallel test execution."""
+        threads = []
+        for index in range(self.parallel):
+            t = threading.Thread(
```

Comment:
Suggestion: It might be easier to use `concurrent.futures.ThreadPoolExecutor` here, rather than pooling yourself.

Reply:
I also agree, but I did want to access the `index`, tbh I didn't research how to do the same with the executor